### PR TITLE
Run serialization tests against upstream types and update v1beta3

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/origin/issues/5483

@deads2k @pmorie @smarterclayton @pweil-

### CODE REVIEW

**Comment by kargakis** [comment]
@pmorie too

**Comment by kargakis** [comment]
[test]

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/serialization_test.go [hunk 1]
```diff
@@ -134,17 +136,19 @@ func TestRoundTripTypes(t *testing.T) {
 	// defer api.Scheme.Log(nil)
 
 	for kind := range api.Scheme.KnownTypes("") {
-		if nonRoundTrippableTypes.Has(kind) {
-			continue
-		}
-		// Try a few times, since runTest uses random values.
-		for i := 0; i < *fuzzIters; i++ {
-			doRoundTripTest(kind, t)
+		for version, codec := range map[string]runtime.Codec{"": api.Codec, "v1beta3": v1beta3.Codec, "v1": v1.Codec} {
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/serialization_test.go]: Given the way that UPSTREAM likes to test by re-running the whole bucket with a different version, I don't think we'll be able to get the structure of this changed moved up.

What about updating the `serialization_test.go` in our `pkg/api` to not exclude kube types?

Comment (kargakis) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/serialization_test.go]: > What about updating the serialization_test.go in our pkg/api to not exclude kube types?

I think this is plausible. Will try it.

Comment (kargakis) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/serialization_test.go]: @deads2k how about now?

**Comment by kargakis** [comment]
Btw, I see that all TypeMeta in v1beta3 need to be updated to unversioned.TypeMeta:)

File: pkg/api/serialization_test.go [hunk 2]
```diff
@@ -405,9 +406,15 @@ func TestSpecificKind(t *testing.T) {
 	}
 }
 
+// Keep this in sync with the respective upstream set
+var nonInternalRoundTrippableTypes = sets.NewString("List", "ListOptions", "PodExecOptions", "PodAttachOptions")
```
Comment (deads2k) [inline pkg/api/serialization_test.go]: Sorry, this isn't obvious to me.  These are somehow broken?  Also, anything that needs to stay in sync should have a unit test to make sure it stays in sync.

Comment (kargakis) [inline pkg/api/serialization_test.go]: It isn't obvious to me either, so I located the PRs which added those upstream:
PodAttachOptions: https://github.com/kubernetes/kubernetes/pull/12119
PodExecOptions: https://github.com/kubernetes/kubernetes/pull/6869
ListOptions: https://github.com/kubernetes/kubernetes/pull/5763
List: https://github.com/kubernetes/kubernetes/pull/2786

@smarterclayton may have a word here

**Comment by kargakis** [comment]
I have removed status which now is an unversioned object and added PodAttachOptions to v1beta3.

Object that still need to be updated in v1beta3:
* PodLogOptions
* Nodes
* PV
* Endpoints
* Services
* PodStatusResult

Based on the massive error returned from the serialization test. We need to make it more readable.

File: pkg/api/serialization_test.go (hunk 2 above)
Comment (pmorie) [inline pkg/api/serialization_test.go]: Should we have test for each of these in `compatibility_test.go`?

**Comment by kargakis** [comment]
Have boiled down serialization failures to Pods

**Comment by kargakis** [comment]
Pod serialization fixed in https://github.com/kargakis/origin/commit/5e64f74aab47234b4967379621cdb093208caf01

this needs review from @pmorie and @pweil-

File: pkg/api/serialization_test.go [hunk 3]
```diff
@@ -415,9 +416,15 @@ func TestSpecificKind(t *testing.T) {
 	}
 }
 
+// Keep this in sync with the respective upstream set
+var nonInternalRoundTrippableTypes = sets.NewString("List", "ListOptions")
+
 func TestTypes(t *testing.T) {
 	for kind, reflectType := range api.Scheme.KnownTypes("") {
-		if !strings.Contains(reflectType.PkgPath(), "/origin/") {
+		if !strings.Contains(reflectType.PkgPath(), "/origin/") && reflectType.PkgPath() != "k8s.io/kubernetes/pkg/api" {
```
Comment (liggitt) [inline pkg/api/serialization_test.go]: why is this needed? if the pkg path does equal `"k8s.io/kubernetes/pkg/api"`, the first condition would already skip it...

**Comment by liggitt** [comment]
can we get a hand crafted test specifically around pod spec in `pkg/api/compatibility_test.go` to make sure an old v1beta3 serialization decodes to what we expect? @pmorie might have a good one

**Comment by kargakis** [comment]
I finally got a clean run locally

File: pkg/api/serialization_test.go (hunk 3 above)
Comment (kargakis) [inline pkg/api/serialization_test.go]: This is AND not OR...

Comment (liggitt) [inline pkg/api/serialization_test.go]: ah. Can we update the name of this test if the intent is to test serialized k8s types?

Comment (kargakis) [inline pkg/api/serialization_test.go]: I like the present name. How about adding godoc?

Comment (liggitt) [inline pkg/api/serialization_test.go]: nevermind, I apparently can't interpret boolean statements today

**Comment by deads2k** [comment]
I think I'm willing to take an issue to figure out this: https://github.com/openshift/origin/pull/5485/files#diff-394377845c8dfd9eec363c059cdde354R420, squash the `UPSTREAM: <carry>'s` down to one and merge to unstick @pmorie's pull.

**Comment by kargakis** [comment]
We still need @pmorie to review https://github.com/kargakis/origin/commit/9a3bcf6c2c96e600d1ad387061cd062000a6f14b

**Comment by kargakis** [comment]
> I think I'm willing to take an issue to figure out this: https://github.com/openshift/origin/pull/5485/files#diff-394377845c8dfd9eec363c059cdde354R420, squash the UPSTREAM: <carry>'s down to one and merge to unstick @pmorie's pull.

Opened https://github.com/openshift/origin/issues/5529

**Comment by deads2k** [comment]
@pmorie I think the last sign off is yours: 

> We still need @pmorie to review kargakis@9a3bcf6

**Comment by pmorie** [comment]
@deads2k I have one comment on the conversion, LGTM otherwise.

**Comment by kargakis** [comment]
re[test]

**Comment by kargakis** [comment]
new-app flake

[test] again

**Comment by deads2k** [comment]
lgtm [merge]

**Comment by pmorie** [comment]
Also LGTM, thanks for the clarification on the conversion.
